packages/@aws-cdk/aws-s3/lib/notifications-resource/lambda-source/src/index.py

```diff
@@ -0,0 +1,89 @@
+import boto3  # type: ignore
+import json
+import urllib.request
+from typing import List
+
+TYPES = [
+    ["TopicConfigurations", "TopicArn"],
+    ["QueueConfigurations", "QueueArn"],
+    ["LambdaFunctionConfigurations", "LambdaFunctionArn"],
+]
+s3 = boto3.client("s3")
+
+
+def handler(event: dict, context):
+    response_status = "SUCCESS"
+    error_message = ""
+    try:
+        props = event["ResourceProperties"]
+        bucket = props["BucketName"]
+        config = prepare_config(
+            current_config=s3.get_bucket_notification_configuration(Bucket=bucket),
+            in_config=props["NotificationConfiguration"],
+            old_config=event.get("OldResourceProperties", {}).get("NotificationConfiguration", {}),
+            request_type=event["RequestType"],
+        )
+        s3.put_bucket_notification_configuration(Bucket=bucket, NotificationConfiguration=config)
+    except Exception as e:
+        print("Failed to put bucket notification configuration:", e)
+        response_status = "FAILED"
+        error_message = f"Error: {str(e)}. "
+    finally:
+        submit_response(event, context, response_status, error_message)
+
+
+def prepare_config(current_config: dict, in_config: dict, old_config: dict, request_type: str) -> dict:
+    config: dict = {}
+    for config_pair in TYPES:
+        config_type, target = config_pair
+        in_configs = in_config.get(config_type, [])
+        old_configs = old_config.get(config_type, [])
+        config[config_type] = find_unmanaged_notifications(current_config, config_type, in_configs, old_configs)
+        if request_type == "Create":
+            config[config_type].extend(in_configs)
```

Comment:
This would still break the upgrade scenario.

Reply:
Maybe some failing tests would help clarify

Reply:
When would `Create` be triggered? On the stack being created or the CustomResource being created? or just by the version of CDK being updated to the new CustomResource and lambda?

If it is the last one, then don't we treat it the same as an `Update`?

Reply:
@iliapolo ?

Reply:
@iliapolo  I feel like the BDD test suite would be the best way to clarify what the solution is.